File: src/Triggers/SetPlayerSpeedTrigger.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;

namespace Celeste.Mod.ProgHelper;

[CustomEntity("progHelper/setPlayerSpeedTrigger")]
public class SetPlayerSpeedTrigger : Trigger {
    private readonly float speedX;
    private readonly float speedY;
    private readonly SetPlayerSpeedMode modeX;
    private readonly SetPlayerSpeedMode modeY;
    private readonly float rangeXMin;
    private readonly float rangeXMax;
    private readonly float rangeYMin;
    private readonly float rangeYMax;

    public SetPlayerSpeedTrigger(EntityData data, Vector2 offset) : base(data, offset) {
        modeX = data.Enum<SetPlayerSpeedMode>("modeX");
        modeY = data.Enum<SetPlayerSpeedMode>("modeY");
        speedX = data.Float("speedX");
        speedY = data.Float("speedY");
        rangeXMin = data.Float("rangeXMin");
        rangeXMax = data.Float("rangeXMax");
        rangeYMin = data.Float("rangeYMin");
        rangeYMax = data.Float("rangeYMax");
    }

    public override void OnEnter(Player player) {
        base.OnEnter(player);

        bool xInRange = player.Speed.X >= rangeXMin && player.Speed.X <= rangeXMax;
        bool yInRange = player.Speed.Y >= rangeYMin && player.Speed.Y <= rangeYMax;

        if (modeX == SetPlayerSpeedMode.Always || modeX == SetPlayerSpeedMode.InRange && xInRange || modeX == SetPlayerSpeedMode.BothInRange && xInRange && yInRange)
            player.Speed.X = speedX;

        if (modeY == SetPlayerSpeedMode.Always || modeY == SetPlayerSpeedMode.InRange && yInRange || modeY == SetPlayerSpeedMode.BothInRange && xInRange && yInRange)
            player.Speed.Y = speedY;
    }
}

File: Loenn/triggers/setPlayerSpeedTrigger.lua
local modeOptions = {
	fieldType = "string",
	options = {
		{ "Always", "Always" },
		{ "Never", "Never" },
		{ "In Range", "InRange" },
		{ "Both in Range", "BothInRange" }
	},
	editable = false
}

return {
	name = "progHelper/setPlayerSpeedTrigger",
	placements = {
		name = "default",
		data = {
			speedX = 0,
			speedY = 0,
			modeX = "Always",
			modeY = "Always",
			rangeXMin = 0,
			rangeXMax = 0,
			rangeYMin = 0,
			rangeYMax = 0
		}
	},
	fieldOrder = {
		"x",
		"y",
		"width",
		"height",
		"speedX",
		"speedY",
		"modeX",
		"modeY",
		"rangeXMin",
		"rangeXMax",
		"rangeYMin",
		"rangeYMax"
	},
	fieldInformation = {
		modeX = modeOptions,
		modeY = modeOptions
	}
}